### PR TITLE
Coverage minima: add more fine-grained control

### PR DESCRIPTION
**Pull request by kitbellew**
Along with existing statement minimum, add branch minimum. Also, include
this pair of control at the package and file level.

### CODE REVIEW

**Comment by suhasgaddam** [comment]
@gslowikowski Can you review this PR?

**Comment by gslowikowski** [comment]
Give me some days

**Comment by kitbellew** [comment]
@gslowikowski when you review this change, do consider the companion https://github.com/scoverage/scoverage-maven-plugin/pull/55
dzięki.

**Comment by kitbellew** [comment]
@gslowikowski friendly reminder to review.

**Comment by gslowikowski** [comment]
@kitbellew I remember about your PRs. I like them very much. It's great that you prepared two pulls for SBT and Maven at the same time.

I've never used JaCoCo before (I used Cobertura for Java). Needed some time to learn JaCoCo and their Gradle, Maven and SBT plugins. I must admit, I like the rules idea:
- in [Gradle](https://docs.gradle.org/current/userguide/jacoco_plugin.html#sec:jacoco_report_violation_rules)
- in [Maven](http://www.jacoco.org/jacoco/trunk/doc/check-mojo.html#rules)

In SBT the [JacocoThresholds](https://github.com/sbt/sbt-jacoco/blob/master/src/main/scala/com/github/sbt/jacoco/report/JacocoThresholds.scala) class is too limited IMO

I'm thinking about next step. Some time ago I was asked about the possibility to set different (higher) coverage minimum values for some stable and well tested packages, files or classes to avoid accidental lowering their coverage.

This is not the topic for your PRs, but I think it's a good idea for future enhancement.

We have to extend Scoverage plugins APIs carefully to be always backward compatible.

I was busy recently, and I didn't want to merge your pulls and then think what to do next. Please, give me some more time. You could think about the more detailed checks I wrote about. Maybe it should be similar to rules in JaCoCo  Maven plugin.

Definitely I don't like SBT configuration. There is only very simple [JacocoThresholds](https://github.com/sbt/sbt-jacoco/blob/master/src/main/scala/com/github/sbt/jacoco/report/JacocoThresholds.scala) class. The field names are not clear for me. E.g. what does `JacocoThresholds.clazz` mean? Percentage of the classes covered or percent of lines/branches/etc. per class?

To sum up, give me some more time. I want to test your proposals and build some projects with JaCoCo for comparison.

**Comment by firehooper** [comment]
How about by glob filepath also like jest? http://jestjs.io/docs/en/configuration.html#coveragethreshold-object .

**Comment by georgeherby** [comment]
Sorry to be that guy but is there any progress on this? As i would love to have branch minimum on my project (cant use forked code on my project) so would love this to be merged

**Comment by kitbellew** [comment]
@gslowikowski are you interested in pursuing this still?

**Comment by mnotti** [comment]
Also hate to be that guy, but would love to have branch minimum, any progress on this PR?

**Comment by ckipp01** [comment]
So I'm currently in the process of getting better acquainted with scoverage @mnotti, so before I review this I need to get a better overarching picture first. I'll then come back and revisit this.

**Comment by kitbellew** [comment]
> So I'm currently in the process of getting better acquainted with scoverage @mnotti, so before I review this I need to get a better overarching picture first. I'll then come back and revisit this.

@ckipp01 didn't realize you've taken over these repositories now... I've been using a locally built maven version for the last 3 years so that definitely works.

**Comment by ckipp01** [comment]
> @ckipp01 didn't realize you've taken over these repositories now...

😬 yes, slowly reviving it.

> I've been using a locally built maven version for the last 3 years so that definitely works.

Ah ok cool, this is great to hear. Are you able to rebase this, and then I'll take a closer look?

**Comment by kitbellew** [comment]
> > @ckipp01 didn't realize you've taken over these repositories now...
> 
> 😬 yes, slowly reviving it.
> 
> > I've been using a locally built maven version for the last 3 years so that definitely works.
> 
> Ah ok cool, this is great to hear. Are you able to rebase this, and then I'll take a closer look?

will do

**Comment by kitbellew** [comment]
@ckipp01 rebased. no changes, no conflicts.

**Comment by kitbellew** [comment]
@ckipp01 should we try rerunning the build? perhaps the error was unrelated to the change.

File: src/sbt-test/scoverage/bad-coverage-file-stmt/build.sbt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+version := "0.1"
+
+scalaVersion := "2.10.4"
```
Comment (ckipp01) [inline src/sbt-test/scoverage/bad-coverage-file-stmt/build.sbt:3]: @kitbellew the test failures are actually due to these build files that have a 2.10 dependency. scoverage is no longer published for 2.10 or 2.11, so you'll need to bump this up. For now stick in the 2.12 range. I need to update all of the test dependencies here since org.specs2 changed names and is not available for 2.13 here either.

Comment (kitbellew) [reply src/sbt-test/scoverage/bad-coverage-file-stmt/build.sbt]: > Also my apologies, but I caused a bunch of conflicts. The specs2 that was in here was really old and I wanted to have a newer framework in there to not block any updates or version or anything. So I went through and updated quite a bit of stuff. If you are able to rebase again, and follow the pattern of the changes, it should be good to go.

no problem. rebased, switched to munit and scala 2.13. ptal.

File: src/main/scala/scoverage/ScoverageKeys.scala [hunk 2]
```diff
@@ -11,7 +11,12 @@ object ScoverageKeys {
   lazy val coverageAggregate = taskKey[Unit]("aggregate reports from subprojects")
   lazy val coverageExcludedPackages = settingKey[String]("regex for excluded packages")
   lazy val coverageExcludedFiles = settingKey[String]("regex for excluded file paths")
-  lazy val coverageMinimum = settingKey[Double]("scoverage-minimum-coverage")
+  lazy val coverageMinimum = settingKey[Double]("scoverage-minimum-coverage-stmt-total")
```
Comment (ckipp01) [inline src/main/scala/scoverage/ScoverageKeys.scala:14]: So digging into this, I've actually never realized that before this was just taking into account statements, and not actual branches. Some of the issues I've seen people report about confusing result not make way more sense since the percentages that are shown are not always logical without realizing this is just the stmt total.

However, this also make me think that we now have a way to set minimal statement coverage and minimal branch coverage. Together, those two will make the up the actual total number we are shooting for. I wonder if that will also cause a bit of confusion since there still may be an assumption that `coverageMinimum` _is_ the total. 🤔 

Part of me wants to actually repurpose `coverageMinimum` to actual make it a total value for averaging stmt and branch, because I do think a lot of people think that's what it means. Then actually move what this to a new `coverageMinimumStmtTotal` key. However, that's sort of breaking behavior, even if people wouldn't see how necessarily.

Just to be super clear, my worry is this. Given the following:

```scala
// build.sbt
coverageMinimum := 56,
coverageFailOnMinimum := true
```

And someone runs their tests and they see:

```
[info] Statement coverage.: 60.00%
[info] Branch coverage....: 50.00%
[info] Coverage reports completed
[info] All done. Coverage was [60.00%]
```

I think it may cause confusion. My mind would go towards, "how is my coverage at 60%"? I have 60% statement and 50% branch meaning a 55% total coverage. This should have failed.

What do you think?

File: src/main/scala/scoverage/ScoverageKeys.scala [hunk 3]
```diff
@@ -11,7 +11,12 @@ object ScoverageKeys {
...
+  lazy val coverageMinimumBranchTotal = settingKey[Double]("scoverage-minimum-coverage-branch-total")
+  lazy val coverageMinimumStmtPerPackage = settingKey[Double]("scoverage-minimum-coverage-stmt-per-package")
+  lazy val coverageMinimumBranchPerPackage = settingKey[Double]("scoverage-minimum-coverage-branch-per-package")
+  lazy val coverageMinimumStmtPerFile = settingKey[Double]("scoverage-minimum-coverage-stmt-per-file")
+  lazy val coverageMinimumBranchPerFile = settingKey[Double]("scoverage-minimum-coverage-branch-per-file")
```
Comment (ckipp01) [inline src/main/scala/scoverage/ScoverageKeys.scala:19]: Also I assumed you were just following suit here  with the `scoverage-minimum-coverage` and the `-`s, but these can just be normal phrases as the descriptions without the `-` separated words.

File: src/main/scala/scoverage/ScoverageKeys.scala [hunk 4]
```diff
@@ -21,4 +26,33 @@ object ScoverageKeys {
...
+  case class CoverageMinimum(
+      statement: Double,
+      branch: Double
+  )
+
+  case class CoverageMinima(
+      total: CoverageMinimum,
+      perPackage: CoverageMinimum,
+      perFile: CoverageMinimum
+  )
```
Comment (ckipp01) [inline src/main/scala/scoverage/ScoverageKeys.scala:56]: Nit, since this file is `ScoverageKeys`, can we keep it _just_ keys and possibly move this out to it's own file?

File: src/main/scala/scoverage/ScoverageSbtPlugin.scala [hunk 5]
```diff
@@ -329,29 +334,72 @@ object ScoverageSbtPlugin extends AutoPlugin {
...
+        checkCoverage(x, s"File:${x.filename}", log, min.perFile)
+      )
 
-    val cper = coverage.statementCoveragePercent
-    val cfmt = coverage.statementCoverageFormatted
+    if (!ok && failOnMin)
+      throw new RuntimeException("Coverage minimum was not reached")
 
+    log.info(
+      s"All done. Coverage was [${coverage.statementCoverageFormatted}%]"
```
Comment (ckipp01) [inline src/main/scala/scoverage/ScoverageSbtPlugin.scala:352]: So depending your thoughts up above, we may need to change this to not just output the statement total but an actual average total, or we might as well output both totals here.

File: src/main/scala/scoverage/ScoverageSbtPlugin.scala [hunk 6]
```diff
@@ -329,29 +334,72 @@ object ScoverageSbtPlugin extends AutoPlugin {
...
+      min: CoverageMinima,
       failOnMin: Boolean
   ): Unit = {
+    val ok: Boolean = checkCoverage(coverage, "Total", log, min.total) &&
+      coverage.packages.forall(x =>
+        checkCoverage(x, s"Package:${x.name}", log, min.perPackage)
+      ) &&
+      coverage.files.forall(x =>
+        checkCoverage(x, s"File:${x.filename}", log, min.perFile)
+      )
```
Comment (ckipp01) [inline src/main/scala/scoverage/ScoverageSbtPlugin.scala:346]: Nit:
```suggestion
      coverage.packages.forall(package =>
        checkCoverage(package, s"Package:${package.name}", log, min.perPackage)
      ) &&
      coverage.files.forall(file =>
        checkCoverage(file, s"File:${file.filename}", log, min.perFile)
      )
```

**Review by ckipp01 — CHANGES_REQUESTED** [review]
Thanks a lot for your work on this @kitbellew. I'm curious to get your opinions on the questions posed.

File: src/main/scala/scoverage/ScoverageKeys.scala (hunk 2 above)
Comment (kitbellew) [reply src/main/scala/scoverage/ScoverageKeys.scala:14]: the main thing i wanted to do is preserve backwards compatibility, so if someone is not interested in these granular keys, their build still keeps working. is there a way to display a deprecation message which says, please use the "total statement" minimum instead of this setting?

File: src/main/scala/scoverage/ScoverageKeys.scala (hunk 3 above)
Comment (kitbellew) [reply src/main/scala/scoverage/ScoverageKeys.scala:19]: i didn't know that :) will change

File: src/main/scala/scoverage/ScoverageKeys.scala (hunk 4 above)
Comment (kitbellew) [reply src/main/scala/scoverage/ScoverageKeys.scala:56]: will do

File: src/main/scala/scoverage/ScoverageSbtPlugin.scala (hunk 6 above)
Comment (kitbellew) [reply src/main/scala/scoverage/ScoverageSbtPlugin.scala:346]: will do

File: src/main/scala/scoverage/ScoverageSbtPlugin.scala (hunk 5 above)
Comment (kitbellew) [reply src/main/scala/scoverage/ScoverageSbtPlugin.scala:352]: will do

File: src/main/scala/scoverage/ScoverageKeys.scala (hunk 2 above)
Comment (ckipp01) [reply src/main/scala/scoverage/ScoverageKeys.scala]: > the main thing i wanted to do is preserve backwards compatibility, so if someone is not interested in these granular keys, their build still keeps working.

Exactly this was my main point as well. Basically the change would force a 2.0 release. Let me think on this a bit. I think ultimately it's the direction that I want to go, but I think I'll want to warn people first. More than likely, if you're alright with it, I'll probably roll back that change on this pr, figure out a way to warn users that the meaning of the key will change in the future, so that there is at least some advance warning. Again, thanks for you thoughts and effort on this.

File: src/main/scala/scoverage/ScoverageSbtPlugin.scala (hunk 6 above)
Comment (ckipp01) [reply src/main/scala/scoverage/ScoverageSbtPlugin.scala]: Ah of course, thanks. 👍🏼

File: src/main/scala/scoverage/ScoverageKeys.scala (hunk 2 above)
Comment (kitbellew) [reply src/main/scala/scoverage/ScoverageKeys.scala]: no problem, whatever you prefer. again, since my understanding of sbt code should be written is rudimentary (i.e., simply parroted the existing code), i will leave it to the experts on how to handle this.

let me know if you want me to make another change. ultimately, i would like to close this out (and especially the maven version which is what we use in our environment).

Comment (kitbellew) [reply src/main/scala/scoverage/ScoverageKeys.scala]: ok, found out how to deprecate a setting. removed the average computation. now it's backwards-compatible but with a warning.

Comment (ckipp01) [reply src/main/scala/scoverage/ScoverageKeys.scala]: Alright, this is perfect. Thanks again for all your work on this and putting up with my back and forths.

**Review by ckipp01 — APPROVED** [review]
LGTM!

Thanks again for taking this across the finish line @kitbellew, especially seeing that you opened two years ago 😆 

I'll start the release process with the compiler plugin in a bit, so if nothing goes wrong, you'll see this somewhat shortly.
